scipy/linalg/tests/test_lapack.py

```diff
@@ -1896,28 +1896,33 @@ def test_gejsv_general(size, dtype, joba, jobu, jobv, jobr, jobt, jobp):
         assert_equal(iwork[2], 0)
 
 @pytest.mark.parametrize("dtype", DTYPES)
-def test_gejsv_specific(dtype):
+def test_gejsv_with_rank_deficient_matrix(dtype):
     """
     Tests more specific cases of routine ?gejsv, specifically for when
-    u and v are not computed, matrix A is 1D, 1x1, empty, lwork = 0,
-    and invalid job? parameters.
+    A is rank deficient. This test calculates and svd of A then keeps only
+    the k singular values, reconstructs a matrix and checks if the
+    decomposition of that matrix contains the correct number of singular
+    values.
     """
     seed(42)
-    atol = 100 * np.finfo(dtype).eps
 
-    m, n = (6, 5)
+    m, n = (6, 5)  # Desired shape of A
+    k = 3  # Desired number of singular values
     gejsv = get_lapack_funcs('gejsv', dtype=dtype)
     A = generate_random_dtype_array((m, n), dtype)
 
     # check that iwork[0] and iwork[1] are the correct rank
     sva, u, v, work, iwork, info = gejsv(A)
-    k = 3
-    sva[np.random.permutation(min(m, n))[k:]] = 0
-    SIGMA = np.diag(work[1] / work[0] * sva[:n])
+
+    # Keep only the first k singular values of A
+    # We then reconstruct A with only the first k singular values
+    # to perform another svd on that matrix.
+    sva[k:] = 0
+    SIGMA = np.diag(work[0] / work[1] * sva[:n])
     A_rank_k = u @ SIGMA @ v.T
     sva, u, v, work, iwork, info = gejsv(A_rank_k)
-    assert_equal(k, iwork[0])
-    assert_equal(k, iwork[1])
+    assert_equal(iwork[0], k)
+    assert_equal(iwork[1], k)
```

Comment:
This is the check that's failing on Azure + Windows for complex128. In this case it's reporting 4 singular values rather than the desired 3. Looking at `sva` it does find 4 singular values, with one being very small:

``sva = array([3.89806589e+00, 1.48313267e+00, 9.06533663e-01, 3.56814239e-16, 0.00000000e+00])``

This only fails for Azure's Windows with complex128, so I'm not sure what's causing the error and how we should treat it.  Could this be treated by choosing a better combination of job? parameters? Or should we simply check that ``iwork[1] = np.count_nonzero(sva)``?

@mdhaber @swallan thoughts?

Reply:
The default `JOBR="R"` (right?) is supposed to zero singular values that are smaller than `sqrt(SFMIN)`, so I would have expected 3.568e-16 to be zeroed (unless `complex128` can actually get higher precision when the complex component is zero?). Anyway, I'm not really concerned by this test failing, but if you're willing, please try the test with JOBA="R" and, if that doesn't work, maybe JOBA="A" (although it sounds like "R" is even more appropriate). If these fail, too, I would skip the test for `complex128`. The user has `sva`, so they can determine which of the numbers are really zero.

Reply:
It looks like some small value(s) occur for any type of joba. I'm going to skip it for now but I'd prefer to play around with it before we eventually merge this as it's passing on my machine / some CIs.